File: frontend-e2e/pkg/flow.ts
/**
 * pkg/flow — runner de "flujos de pasos" autodocumentados para el frontend-e2e.
 *
 * Es el espejo en TypeScript de backend-e2e/pkg/flow/flow.go: la consola muestra
 * el paso a paso igual que el backend. Cada flujo es una secuencia de pasos
 * {título, descripción, acción}; el runner:
 *   - numera los pasos automáticamente ([i/N], sin totales hardcodeados),
 *   - imprime QUÉ hace cada paso (la descripción, `↳`) antes de ejecutarlo,
 *   - ejecuta la acción y muestra el resultado (✓ detalle / ✗ error),
 *   - se detiene en el primer paso que falle (run) o los corre todos (runAll), y
 *   - imprime un resumen final (N/N pasos · tiempo).
 *
 * Cada paso se envuelve en `test.step(...)`, así que ADEMÁS aparece agrupado en
 * el HTML report y el trace de Playwright (algo que el backend no tiene).
 *
 * Salida limpia y paralelismo: el `playwright.config.ts` corre con
 * `fullyParallel: true`, así que si varios flujos imprimen a la vez sus líneas se
 * entremezclan. Para la narración paso-a-paso legible (igual que el CLI serial del
 * backend) usá `npm run e2e:steps`, que fuerza `--workers=1`.
 *
 * Los datos que un paso necesita de otro (p.ej. el loanRequestId que produce el
 * submit) viajan por el `Ctx`. Las dependencias estáticas (page, hashes) las
 * captura cada closure de paso.
 */
import { test } from '@playwright/test';

// Colores ANSI — mismos códigos que backend-e2e/pkg/client/client.go.
// Se desactivan fuera de TTY o con FORCE_COLOR=0 para no ensuciar logs de CI.
const useColor = process.env.FORCE_COLOR !== '0' && (!!process.stdout.isTTY || !!process.env.FORCE_COLOR);
const c = (code: string, s: string) => (useColor ? `\x1b[${code}m${s}\x1b[0m` : s);
const blue = (s: string) => c('34', s);
const green = (s: string) => c('32', s);
const red = (s: string) => c('31', s);
const cyan = (s: string) => c('36', s);
const gray = (s: string) => c('90', s);

/** Ctx transporta valores entre pasos del mismo flujo (espejo de flow.Ctx). */
export class Ctx {
    private vals = new Map<string, unknown>();
    set(k: string, v: unknown): void {
        this.vals.set(k, v);
    }
    get(k: string): unknown {
        return this.vals.get(k);
    }
    str(k: string): string {
        const v = this.vals.get(k);
        return typeof v === 'string' ? v : '';
    }
    int(k: string): number {
        const v = this.vals.get(k);
        return typeof v === 'number' ? v : Number(v) || 0;
    }
}

/**
 * Acción de un paso. Devuelve un detalle corto del resultado (se imprime junto
 * al ✓); `void`/`undefined` => no imprime detalle. Lanza para fallar el paso.
 */
export type StepRun = (ctx: Ctx) => Promise<string | void> | string | void;

interface Step {
    title: string;
    desc: string;
    run: StepRun;
}

const banner = (t: string): void => {
    const bar = '='.repeat(70);
    console.log(cyan(`${bar}\n ${t} \n${bar}`));
};
const printStep = (i: number, total: number, title: string): void =>
    console.log(blue(`\n[${i}/${total}] ${title}`));
const printRule = (rule: string): void => console.log(`  ↳ ${gray(rule)}`);
const printOK = (msg: string): void => console.log(green(`  ✓ ${msg}`));
const printFail = (msg: string): void => console.log(red(`  ✗ ${msg}`));
const secs = (startMs: number): string => `${((Date.now() - startMs) / 1000).toFixed(1)}s`;

/** Flow es una secuencia de pasos autodocumentada (espejo de flow.Flow). */
export class Flow {
    private steps: Step[] = [];
    constructor(
        private name: string,
        private summary = '',
    ) {}

    /** Agrega un paso (encadenable). */
    step(title: string, desc: string, run: StepRun): this {
        this.steps.push({ title, desc, run });
        return this;
    }

    /** Cantidad de pasos. */
    get length(): number {
        return this.steps.length;
    }

    private header(): void {
        banner(this.name);
        if (this.summary) console.log(`  ${gray(this.summary)}`);
    }

    /** Imprime el paso a paso DOCUMENTADO sin ejecutar ninguna acción (espejo de Explain). */
    explain(): void {
        this.header();
        this.steps.forEach((s, i) => {
            printStep(i + 1, this.steps.length, s.title);
            if (s.desc) printRule(s.desc);
        });
        console.log(gray('\n(explain: documentación del flujo; no se ejecutó nada)'));
    }

    /** Ejecuta el flujo; se detiene y lanza en el primer paso que falle. */
    async run(ctx: Ctx = new Ctx()): Promise<void> {
        this.header();
        const start = Date.now();
        for (let i = 0; i < this.steps.length; i++) {
            const s = this.steps[i];
            printStep(i + 1, this.steps.length, s.title);
            if (s.desc) printRule(s.desc);
            try {
                const detail = await test.step(s.title, () => s.run(ctx));
                if (detail) printOK(detail);
            } catch (err) {
                printFail(`${s.title}: ${(err as Error).message}`);
                console.log(
                    red(`\n🔴 FLUJO DETENIDO en el paso ${i + 1}/${this.steps.length} (${s.title}) · ${secs(start)}`),
                );
                throw err;
            }
        }
        console.log(green(`\n🟢 FLUJO OK · ${this.steps.length}/${this.steps.length} pasos · ${secs(start)}`));
    }

    /**
     * Ejecuta TODOS los pasos SIN detenerse en el primer fallo (para baterías de
     * aserciones independientes, p.ej. rutas negativas). Imprime ✓/✗ por paso y un
     * resumen N/M; lanza al final si alguno falló (espejo de RunAll).
     */
    async runAll(ctx: Ctx = new Ctx()): Promise<void> {
        this.header();
        const start = Date.now();
        let failed = 0;
        for (let i = 0; i < this.steps.length; i++) {
            const s = this.steps[i];
            printStep(i + 1, this.steps.length, s.title);
            if (s.desc) printRule(s.desc);
            try {
                const detail = await test.step(s.title, () => s.run(ctx));
                if (detail) printOK(detail);
            } catch (err) {
                failed++;
                printFail(`${s.title}: ${(err as Error).message}`);
            }
        }
        const ok = this.steps.length - failed;
        if (failed > 0) {
            console.log(red(`\n🔴 ${ok}/${this.steps.length} pasos OK · ${failed} fallaron · ${secs(start)}`));
            throw new Error(`${failed}/${this.steps.length} pasos fallaron`);
        }
        console.log(green(`\n🟢 FLUJO OK · ${ok}/${this.steps.length} pasos · ${secs(start)}`));
    }
}

File: backend-e2e/pkg/flow/flow.go
// Package flow es un runner de "flujos de pasos" autodocumentados para el harness E2E.
//
// Cada flujo es una secuencia de pasos {Título, Descripción, Acción}. El runner:
//   - numera los pasos automáticamente (sin totales hardcodeados),
//   - imprime QUÉ hace cada paso (la descripción) antes de ejecutarlo,
//   - ejecuta la acción y muestra el resultado (✓ detalle / ✗ error),
//   - se detiene en el primer paso que falle, y
//   - imprime un resumen final (N/N pasos · tiempo).
//
// Con Explain() imprime SOLO la documentación del flujo (sin ejecutar nada): sirve para entender
// el paso a paso de un comando sin tocar el backend.
//
// Los datos que un paso necesita de otro (p.ej. el user_request_id de la entrada que usa el cierre)
// viajan por el Ctx. Las dependencias estáticas (db, cfg) las captura cada closure de paso.
package flow

import (
	"fmt"
	"time"

	"creditop-tests/pkg/client"
)

// Ctx transporta valores entre pasos del mismo flujo.
type Ctx struct{ vals map[string]interface{} }

// NewCtx crea un contexto vacío (para correr pasos embebidos con RunInline).
func NewCtx() *Ctx { return &Ctx{vals: map[string]interface{}{}} }

// Set guarda un valor para que lo lean pasos posteriores.
func (c *Ctx) Set(k string, v interface{}) { c.vals[k] = v }

// Get devuelve el valor crudo (nil si no existe).
func (c *Ctx) Get(k string) interface{} { return c.vals[k] }

// Int64 devuelve el valor como int64 (0 si no existe o no es numérico).
func (c *Ctx) Int64(k string) int64 {
	switch v := c.vals[k].(type) {
	case int64:
		return v
	case int:
		return int64(v)
	case float64:
		return int64(v)
	}
	return 0
}

// Int devuelve el valor como int.
func (c *Ctx) Int(k string) int { return int(c.Int64(k)) }

// Str devuelve el valor como string ("" si no existe).
func (c *Ctx) Str(k string) string { s, _ := c.vals[k].(string); return s }

// Step es un paso del flujo: un título (qué), una descripción (qué hace / por qué) y la acción.
// Run devuelve un detalle corto del resultado (se imprime junto al ✓) y un error si el paso falla.
type Step struct {
	Title string
	Desc  string
	Run   func(c *Ctx) (string, error)
}

// Flow es una secuencia de pasos autodocumentada.
type Flow struct {
	Name    string
	Summary string
	steps   []Step
}

// New crea un flujo con nombre (cabecera) y un resumen de una línea.
func New(name, summary string) *Flow { return &Flow{Name: name, Summary: summary} }

// Step agrega un paso (encadenable).
func (f *Flow) Step(title, desc string, run func(*Ctx) (string, error)) *Flow {
	f.steps = append(f.steps, Step{Title: title, Desc: desc, Run: run})
	return f
}

// Add agrega pasos ya construidos (encadenable). Útil para componer entrada + comercio + cierre.
func (f *Flow) Add(steps ...Step) *Flow { f.steps = append(f.steps, steps...); return f }

// Len es la cantidad de pasos.
func (f *Flow) Len() int { return len(f.steps) }

func (f *Flow) header() {
	client.Banner(f.Name)
	if f.Summary != "" {
		fmt.Printf("  %s%s%s\n", client.CGray, f.Summary, client.CReset)
	}
}

func (f *Flow) printStep(i int, s Step) {
	client.PrintStep(i+1, len(f.steps), s.Title)
	if s.Desc != "" {
		client.PrintRule(s.Desc)
	}
}

// Explain imprime el paso a paso DOCUMENTADO del flujo sin ejecutar ninguna acción.
func (f *Flow) Explain() {
	f.header()
	for i, s := range f.steps {
		f.printStep(i, s)
	}
	fmt.Printf("\n%s(--explain: documentación del flujo; no se ejecutó nada)%s\n", client.CGray, client.CReset)
}

// Run ejecuta el flujo con cabecera + resumen. Devuelve error en el primer paso que falle.
func (f *Flow) Run() error {
	f.header()
	start := time.Now()
	c := NewCtx()
	if err := f.exec(c, start); err != nil {
		return err
	}
	fmt.Printf("\n%s🟢 FLUJO OK · %d/%d pasos · %s%s\n", client.CGreen, len(f.steps), len(f.steps), since(start), client.CReset)
	return nil
}

// RunInline ejecuta los pasos sobre un Ctx dado, SIN cabecera ni resumen (para embeber un sub-flujo,
// p.ej. la entrada del canal cuando el comando trae su propio encabezado). Devuelve el primer error.
func (f *Flow) RunInline(c *Ctx) error { return f.exec(c, time.Now()) }

// RunAll ejecuta TODOS los pasos SIN detenerse en el primer fallo (para baterías de aserciones
// independientes, p.ej. rutas negativas/anti-fraude). Imprime ✓/✗ por paso y un resumen N/M;
// devuelve error si alguno falló.
func (f *Flow) RunAll() error {
	f.header()
	start := time.Now()
	c := NewCtx()
	failed := 0
	for i, s := range f.steps {
		f.printStep(i, s)
		detail, err := s.Run(c)
		if err != nil {
			failed++
			client.PrintFail(fmt.Sprintf("%s: %v", s.Title, err))
			continue
		}
		if detail != "" {
			client.PrintOK(detail)
		}
	}
	ok := len(f.steps) - failed
	if failed > 0 {
		fmt.Printf("\n%s🔴 %d/%d pasos OK · %d fallaron · %s%s\n", client.CRed, ok, len(f.steps), failed, since(start), client.CReset)
		return fmt.Errorf("%d/%d pasos fallaron", failed, len(f.steps))
	}
	fmt.Printf("\n%s🟢 FLUJO OK · %d/%d pasos · %s%s\n", client.CGreen, ok, len(f.steps), since(start), client.CReset)
	return nil
}

func (f *Flow) exec(c *Ctx, start time.Time) error {
	for i, s := range f.steps {
		f.printStep(i, s)
		detail, err := s.Run(c)
		if err != nil {
			client.PrintFail(fmt.Sprintf("%s: %v", s.Title, err))
			fmt.Printf("\n%s🔴 FLUJO DETENIDO en el paso %d/%d (%s) · %s%s\n",
				client.CRed, i+1, len(f.steps), s.Title, since(start), client.CReset)
			return err
		}
		if detail != "" {
			client.PrintOK(detail)
		}
	}
	return nil
}

func since(start time.Time) string { return fmt.Sprintf("%.1fs", time.Since(start).Seconds()) }
